persist/sqldb/workflow_archive.go

```diff
@@ -70,9 +70,15 @@ func (r *workflowArchive) ArchiveWorkflow(wf *wfv1.Workflow) error {
 		return err
 	}
 	return r.session.Tx(context.Background(), func(sess sqlbuilder.Tx) error {
-		// We assume that we're much more likely to be inserting rows that updating them, so we try and insert,
-		// and if that fails, then we update.
-		// There is no check for race condition here, last writer wins.
+		_, err := sess.
+			DeleteFrom(archiveTableName).
+			Where(db.Cond{"clustername": r.clusterName}).
+			And(db.Cond{"instanceid": r.instanceID}).
```

Comment:
`instanceid` is not needed here, it's not part of the PK, and we don't have an index matching (clustername, instanceid, uid).

The indexes we have are:
1. PK combo
2. (clustername, instanceid, namespace)

Reply:
you're correct, the instance ID will act as a filter - I think we should keep this because it'll prevent one instance from delete workflows from another instance - thoughts

Reply:
Assume there is a case like you mentioned, it doesn't solve your problem - record not getting deleted, you will still get dup_key error.

Reply:
I can't see when this will happen - you'd have needed to have re-configure your controller between invocations...?

Reply:
This only happens when:
1. A workflow object already exists in the cluster with `instanceid01`, and archived with same instanceid;
2. Re-configure your controller to serve `instanceid02`, or there's another controller in the same cluster serving `instanceid02`;
3. Update the object's label to mark it as `instanceid02`;
4. retry - archive

Reply:
If we consider this kind of cases, we might need `instanceid` as part of the PK. :)

Reply:
What happens if we assume instance ID never changes?

Reply:
if wf object instance ID label never changes, no matter how controller config changes, there won't be a case like that.

Reply:
ok, so should I be  allowed to ask a controller to delete a workflow that belongs to another controller?

Reply:
Well, if wf object instance ID label never changes, you will not experience the situation that there is a record with same (clustername, uid) but different instanceid existing in the DB;

If wf object instant ID label is changed, which means the current wf object you are working on, was already archived by another controller, and there's a DB record existing. In this case,  if you add `instanceid` condition in your delete sql, you will not delete the record, and you still get `duplicate key` error when you insert.

Reply:
ok, removed - can you please review?